Bluetooth: Skip address setting for passive scan

trond-snekvik opened the pull request:
The passive scanner isn't actually using the advertising address, and
doesn't need to set it during scan initialization.

set_random_addr always fails if there's an active advertiser, which
would cause needless complications for the Bluetooth Mesh LPN case.

Fixes #22088.

Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

trond-snekvik committed:
Bluetooth: host: Skip set passive scan when not scanning with identity

When privacy is disabled by default the scanner still protects it's
identity with the use of NRPA addresses. We should not set the identity
address for the passive scanner unless the Kconfig option to scan with
the identity has been enabled.

This makes passive scanner behave the same way as an active scanner
since none of them will report directed advertising reports towards
the scanners identity.

This also enables the advertiser to switch out the random address which
is needed for the Bluetooth Mesh LPN case.

Fixes #22088.

Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

joerchan commented:
This is a continuation of https://github.com/zephyrproject-rtos/zephyr/pull/21714

jhedberg commented:
I might remember wrong, but I have a faint memory of us discussing this with the controller guys some time ago, and there was something that still justified setting the address for passive scanning. Perhaps it was directed advertising, or something similar, or then I just remember wrong and it was for something else. @cvinayak @carlescufi any comments?

joerchan commented:
> I might remember wrong, but I have a faint memory of us discussing this with the controller guys some time ago, and there was something that still justified setting the address for passive scanning. Perhaps it was directed advertising, or something similar, or then I just remember wrong and it was for something else. @cvinayak @carlescufi any comments?

@jhedberg That's true. But then using advertiser with NRPA at the same time would not work. We could possibly set random address if the advertiser is not running. Otherwise we have a structural conflict with the passive scanner on using the random address as well.

joerchan commented:
@jhedberg @carlescufi @trond-snekvik 
Maybe we could try to set the random address, but if it fails we continue without it (for passive scanner only). Then we would not get directed advertising reports. Then we could support the Mesh use-case of NRPA advertiser and Passive scanner.

Of course this should be documented clearly. It would only affect users who start advertiser with NRPA or an identity different than the default identity (BT_ID_DEFAULT).

joerchan commented:
Furthermore. This should be documented, or possibly handled together with the `Central Address Resolution` characteristic.

joerchan commented:
Current behaviour when Privacy is OFF and Random Static Identity

 - Active scanner sets an NRPA if not advertising or Kconfig option BT_SCAN_WITH_IDENTITY is set.
   Active scanner currently will not receive directed advertise reports.

 - Passive scanner sets random address.
   Passive scanner currently can received directed advertise reports.

 - Passive scanner with non-connectable advertiser
   Advertiser will overwrite random address of passive scanner, and will not receive directed advertise reports anymore.

 - Active scanner with non-connectable advertiser
   Both will try to set NRPA, and it will fail (because it is not the same NRPA)

Suggested action:
 - Switch default BT_SCAN_WITH_IDENTITY to on. (off by default) to support directed advertising reports.
 - Mesh sets BT_SCAN_WITH_IDENTITY to off
 - When BT_SCAN_WITH_IDENTITY is off, allow advertiser to overwrite the address NRPA of the scanner.
   - Should passive scanner set NRPA instead of Random Static Identity?
     It would be confusing that passive scanner receives directed advertising reports, but not active scanner

carlescufi commented:
Conclusion after meeting:

- `BT_SCAN_WITH_IDENTITY` enables not only using (and setting) the identity address when scanning (and no privacy) but also the reception of directed adv reports for **both** active and passive scanning.
- Document this fact in both the Kconfig option help text and the `bt_le_scan_start()` doxygen

joerchan committed:
Bluetooth: host: Drop directed adv reports for NRPA passive scanner

Drop directed advertiser reports when the passive scanner should have
been using the NRPA address. The advertiser has overwritten the NRPA
with it's identity address instead and a peer is sending directed
advertiser packets to it.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Prevent adv overwriting passive scan identity address

Prevent the advertiser from overwriting the passive scanners identity
address when the scanner has been configured to scan using the identity.
In this case the LE Set Random Address command would not prevent the
address from being overwritten. So instead we explicitly stop it in the
host.

Also fix original code function not working at all since the first
if statement was should had a bad check. Resulting in it always
returning success.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Document privacy-disabled directed advertising reports

Document why a privacy-disabled scanner will not notify about directed
advertising reports. This is the default behaviour of the
privacy-disabled scanner. In order to receive the reports the option
BT_SCAN_WITH_IDENTITY must be enabled.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Scanner must be disabled to update to NRPA address

When advertiser is disabled we should update the random address for both
passive and active scanner back to an NRPA. But this command will fail
because the if the scanner is an active scanner, we must disable and
re-enabled the active scanner after setting the random address.
This behavior should not be there when scanner is configured to scan
with identity address.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Refresh advertiser NRPA when active scan is using NRPA

When starting a non-connectable advertiser and an active scanner is also
using NRPA address then this use-case should be supported. A new
advertiser that is non-connectable should have a fresh NRPA address
every time it is started, so we must refresh the NRPA used by the active
scanner.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

trond-snekvik commented:
The mesh still works after @joerchan's changes.

joerchan committed:
Bluetooth: tests: Update hci_prop_evt for LE Set Random Address command

Accept LE Set Random Address command during HCI init.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

cvinayak commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -695,6 +690,32 @@ static bool bt_le_adv_random_addr_check(const struct bt_le_adv_param *param)
+		}
+
+
+		if (((param->options & BT_LE_ADV_OPT_CONNECTABLE) ||
```
`((param->options & BT_LE_ADV_OPT_CONNECTABLE) || !(param->options & BT_LE_ADV_OPT_USE_IDENTITY))`

Is redundant as it evaluates to NOT operation of line 703, instead use  `else if` clause.

joerchan replied:
Thanks. Updated.
